src/observers.ts

```diff
@@ -0,0 +1,243 @@
+    const totalEpochsPrescribed =
+      distributions.observers[address]?.totalEpochsPrescribedCount || 0;
+    const totalEpochsSubmitted =
+      distributions.observers[address]?.submittedEpochCount || 0;
+    // defaults to one again, encouraging new gateways to join
+    const observerRewardRatioWeight = totalEpochsPrescribed
+      ? totalEpochsSubmitted / totalEpochsPrescribed
+      : 1;
+
+    // calculate composite weight based on sub weights
+    const compositeWeight =
+      stakeWeight *
+      gatewayTenureWeight *
+      gatewayRewardRatioWeight *
+      observerRewardRatioWeight;
+
+    weightedObservers.push({
+      gatewayAddress: address,
+      observerAddress: eligibleGateway.observerWallet,
+      stake,
+      start: eligibleGateway.start,
+      stakeWeight,
+      tenureWeight: gatewayTenureWeight,
+      gatewayRewardRatioWeight,
+      observerRewardRatioWeight,
+      compositeWeight,
+      normalizedCompositeWeight: compositeWeight,
+    });
+    // total weight for all eligible gateways
+    totalCompositeWeight += compositeWeight;
+  }
+
+  // calculate the normalized composite weight for each observer - do not default to one as these are dependent on the total weights of all observers
+  for (const weightedObserver of weightedObservers) {
+    weightedObserver.normalizedCompositeWeight = totalCompositeWeight
+      ? weightedObserver.compositeWeight / totalCompositeWeight
+      : 0;
+  }
+
+  // return all the observers if there are fewer than the number of observers per epoch
+  if (MAXIMUM_OBSERVERS_PER_EPOCH >= Object.keys(weightedObservers).length) {
+    return weightedObservers;
+  }
+
+  // deterministic way to get observers per epoch based on the epochs end height
+  const blockHeightEntropyHash = await getEntropyHashForEpoch({
+    epochStartHeight,
+  });
+
+  // note: this should always result to MAXIMUM_OBSERVERS_PER_EPOCH
+  const prescribedObservers: Set<WeightedObserver> = new Set();
+  let hash = blockHeightEntropyHash; // our starting hash
+  for (let i = 0; i < MAXIMUM_OBSERVERS_PER_EPOCH; i++) {
+    const random = hash.readUInt32BE(0) / 0xffffffff; // Convert hash to a value between 0 and 1
+    let cumulativeNormalizedCompositeWeight = 0;
+    for (const observer of weightedObservers) {
+      // skip observers that have already been prescribed
+      if (prescribedObservers.has(observer)) continue;
+      // add the observers normalized composite weight to the cumulative weight
+      cumulativeNormalizedCompositeWeight += observer.normalizedCompositeWeight;
+      // if the random value is less than the cumulative weight, we have found our observer
+      if (random <= cumulativeNormalizedCompositeWeight) {
+        prescribedObservers.add(observer);
+        break;
+      }
+      // Compute the next hash for the next iteration
+      hash = await SmartWeave.arweave.crypto.hash(hash, 'SHA-256');
+    }
+  }
```

Comment:
@arielmelendez  - can you look through this? i don't like the complexity here so any thoughts welcom